### PR TITLE
feat: add support for public IP connections

### PR DESCRIPTION
**Pull request by jackwotherspoon**
Add support for `--public-ip` flag and public IP query option to allow connections to AlloyDB over public IP.

- [X] Add support for flag and query params
- [x] Get current tests to pass
- [x] Add integration test for public IP
- [x] Update README with sample

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -3,7 +3,7 @@ module github.com/GoogleCloudPlatform/alloydb-auth-proxy
 go 1.20
 
 require (
-	cloud.google.com/go/alloydbconn v1.5.2
+	cloud.google.com/go/alloydbconn v1.5.3-0.20240118193619-b19b9641fbe1
```
Comment (jackwotherspoon) [inline go.mod:6]: Pinned to AlloyDB Go Connector public IP PR most recent commit. 

Remove after merging Go Connector dep bump

File: internal/proxy/proxy.go [hunk 2]
```diff
@@ -181,6 +189,19 @@ type Config struct {
...
+// dialOptions interprets appropriate dial options for a particular instance
+// configuration
+func dialOptions(c Config, i InstanceConnConfig) []alloydbconn.DialOption {
+	var opts []alloydbconn.DialOption
+
+	// If public IP is enabled at the instance level, or public IP is enabled
+	// globally, add the option.
+	if i.PublicIP || c.PublicIP {
+		opts = append(opts, alloydbconn.WithPublicIP())
+	}
```
Comment (jackwotherspoon) [inline internal/proxy/proxy.go:201]: This logic may need to be updated, depends on intended behavior. If someone has globally set `publicIP` to True  but query params it with false do we want the query param to override?

File: internal/proxy/proxy.go [hunk 3]
```diff
@@ -60,6 +60,10 @@ type InstanceConnConfig struct {
 	// AutoIAMAuthN enables automatic IAM authentication on the instance only.
 	// See Config.AutoIAMAuthN for more details.
 	AutoIAMAuthN bool
+
+	// PublicIP tells the proxy to attempt to connect to the db instance's
+	// public IP address instead of the private IP address
+	PublicIP bool
```
Comment (jackwotherspoon) [inline internal/proxy/proxy.go:66]: should I make this `PublicIP *bool` as it is in the Cloud SQL Proxy?

File: internal/proxy/proxy.go (hunk 2 above)
Comment (enocom) [reply internal/proxy/proxy.go:201]: Yes we do want the param to override in that case. Think connecting to three instances, all but one is public ip. You'd set the global flag and disable the flag with a query param for the private instance.

File: internal/proxy/proxy.go (hunk 3 above)
Comment (enocom) [reply internal/proxy/proxy.go:66]: A bool is fine. If we need to change it in the future, this is a private interface so that's fine.

**Review by enocom — COMMENTED** [review]
PR looks good. We just need to make sure query params can override the global option.

File: internal/proxy/proxy.go (hunk 3 above)
Comment (enocom) [reply internal/proxy/proxy.go:66]: Scratch that. We need a bool pointer to support this:

- user sets public IP to false globally
- user enables public ip on a per-instance level.

File: internal/proxy/proxy.go (hunk 2 above)
Comment (jackwotherspoon) [reply internal/proxy/proxy.go]: Should be able to override now, added test to make sure

File: internal/proxy/proxy.go (hunk 3 above)
Comment (jackwotherspoon) [reply internal/proxy/proxy.go]: i added a test case for the opposite of this as i think it is more likely, 

- public IP enabled globally
- user disables public IP on a per-instance level

**Review by enocom — APPROVED** [review]

**Pull request merged**